integration_tests/DATA_GEN.md

```diff
@@ -0,0 +1,402 @@
+around this.
+
+### DistinctDistribution
+
+The `DistinctDistribution` generates seeds from min seed to max seed, but
+walks them in a non-sequential order. The data generation functions also maps
+the data in a pseudo-random way so the unique keys will not show this very
+clearly.
+
+```scala
+val dataTable = DBGen().addTable("data", "a byte", 10)
+dataTable("a").setSeedMapping(DistinctDistribution())
+dataTable.toDF(spark).show
++---+
+|  a|
++---+
+| 97|
+|121|
+|-64|
+| 51|
+|-66|
+| 83|
+| 90|
+| 76|
+| 48|
+| 80|
++---+
+```
+
+Be careful with this. If you set the seed range too small to support all
+the rows, or if the cardinality of the data itself is too small, like a byte,
+you will get repeated values.
+
+But a `DistinctDistribution` is really helpful when wanting to generate a
+fact table for a join where you want the primary key to be distinct.
+
+Please note that this does not work well with structs or other nested types
+because the seeds passed into the parents do not control what the children
+generate. See [Multi-Column Keys](#multi-column-keys) for info on how to work
+around this.
+
+## Multi-Column Keys
+
+With the basic tools provided we can now replicate a lot of processing. We can do
+complicated things like a join with a fact table followed by an aggregation. 
+
+```scala
+val dbgen = DBGen()
+val factTable = dbgen.addTable("facts", "join_key long, value int", 1000L)
+factTable("join_key").setSeedRange(0, 999).setSeedMapping(DistinctDistribution())
+val dataTable = dbgen.addTable("data", "join_key long, agg_key long", 100000000L)
+dataTable("join_key").setSeedRange(0, 999)
+dataTable("agg_key").setSeedRange(0, 9)
+val fdf = factTable.toDF(spark)
+val ddf = dataTable.toDF(spark)
+spark.time(fdf.join(ddf).groupBy("agg_key").agg(min("value"), 
+  max("value"), sum("value")).orderBy("agg_key").show())
++--------------------+-----------+----------+-------------------+
+|             agg_key| min(value)|max(value)|         sum(value)|
++--------------------+-----------+----------+-------------------+
+|-5395130417820713410|-2146262091|2137753315|-790927616093345360|
+|-2676343845813715061|-2146262091|2137753315|-790805185346631680|
+|-2519888240660072622|-2146262091|2137753315|-790643842502125280|
+| -423375629809417248|-2146262091|2137753315|-790620036523597620|
+| 2201036318770978408|-2146262091|2137753315|-790864265299987700|
+| 2236810499658231116|-2146262091|2137753315|-791190430932705540|
+| 5462872210444757997|-2146262091|2137753315|-791094653391187280|
+| 5615780013263136643|-2146262091|2137753315|-790580412619137960|
+| 6937819975370740711|-2146262091|2137753315|-791081287243774740|
+| 8365571621216757512|-2146262091|2137753315|-791155236047506840|
++--------------------+-----------+----------+-------------------+
+
+Time taken: 890163 ms
+```
+
+Or you could run it with the RAPIDS Accelerator where the data generations is still
+done on the CPU and cut the time down to just 75,696 ms
```

Comment:
75,696 ms on which hardware?

Reply:
My desktop. It is more of a job at CPUs than anything.